src/backend/distributed/planner/modify_planner.c

```diff
@@ -641,3 +644,315 @@ ExtractPartitionValue(Query *query, Var *partitionColumn)
+	}
+
+	shardCount = ShardIntervalCount(distributedTableId);
+
+	if (shardCount == 0)
+	{
+		return false;
+	}
+
+	return true;
+}
+
+
+/* MultiRouterPlanCreate creates a physical plan for given router plannable query. */
+MultiPlan *
+MultiRouterPlanCreate(Query *query)
+{
+	Task *task = NULL;
+	Job *job = NULL;
+	MultiPlan *multiPlan = NULL;
+
+	task = SingleShardSelectTask(query);
+
+	job = SingleShardSelectJob(query,task);
+
+	multiPlan = CitusMakeNode(MultiPlan);
+	multiPlan->workerJob = job;
+	multiPlan->masterQuery = NULL;
+	multiPlan->masterTableName = NULL;
+
+	return multiPlan;
+}
+
+
+/* SingleShardSelectTask builds a Task to represent a single shard select query */
+static Task *
+SingleShardSelectTask(Query *query)
+{
+	Task *task = NULL;
+	ShardInterval *shardInterval = SingleShardSelectShardInterval(query);
+	StringInfo queryString = makeStringInfo();
+	uint64 shardId = INVALID_SHARD_ID;
+	bool upsertQuery = false;
+	CmdType commandType = query->commandType;
+	FromExpr *joinTree = NULL;
+
+	Assert(shardInterval != NULL);
+	Assert(commandType == CMD_SELECT);
+
+	shardId = shardInterval->shardId;
+
+	/*
+	 * Convert the qualifiers to an explicitly and'd clause, which is needed
+	 * before we deparse the query.
+	 */
+	joinTree = query->jointree;
+	if ((joinTree != NULL) && (joinTree->quals != NULL))
+	{
+		Node *whereClause = joinTree->quals;
+		if (IsA(whereClause, List))
+		{
+			joinTree->quals = (Node *) make_ands_explicit((List *) whereClause);
+		}
+	}
+
+	deparse_shard_query(query, shardInterval->relationId, shardId, queryString);
+	ereport(DEBUG4, (errmsg("distributed statement: %s", queryString->data)));
+
+	task = CitusMakeNode(Task);
+	task->jobId = INVALID_JOB_ID;
+	task->taskId = INVALID_TASK_ID;
+	task->taskType = ROUTER_TASK;
+	task->queryString = queryString->data;
+	task->anchorShardId = shardId;
+	task->dependedTaskList = NIL;
+	task->upsertQuery = upsertQuery;
+
+	return task;
+}
+
+
+/*
+ * SingleShardSelectShardInterval determines the single shard targeted by a
+ * provided distributed modification command. If no matching shards exist, or
+ * if the modification targets more than one one shard, this function raises
+ * an error.
+ */
+static ShardInterval *
+SingleShardSelectShardInterval(Query *query)
+{
+	List *restrictClauseList = NIL;
+	List *prunedShardList = NIL;
+	Index tableId = 1;
+
+	Oid distributedTableId = ExtractFirstDistributedTableId(query);
+	List *shardIntervalList = NIL;
+
+	/* error out if no shards exist for the table */
+	shardIntervalList = LoadShardIntervalList(distributedTableId);
+	if (shardIntervalList == NIL)
+	{
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("could not find any shards for this table")));
+	}
+
+	restrictClauseList = QueryRestrictList(query);
+	prunedShardList = PruneShardList(distributedTableId, tableId, restrictClauseList,
+									 shardIntervalList);
+
+	Assert(list_length(prunedShardList) <= 1);
+
+	if (list_length(prunedShardList) > 1)
```

Comment:
It could be 0 if all shards are pruned away. 

We later introduced a fix
1 - there has to be some shards
2 - where clause can reference partition column once.

However, there is a slight change here with the following scenario
One creates hash partitioned distributed table, creates shards on it.  But removes one of them after creation. This results in having a missing hash range.

There is probability that query can address that particular shard, and after shard pruning we are left with no shards.